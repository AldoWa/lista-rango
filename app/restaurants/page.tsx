
import { Metadata } from 'next';
import { CardRestaurant } from '@/app/ui/card-restaurant';
import { getRestaurants } from '@/app/libs/requests';
import { isValidHours } from '@/app/libs/utils';
import { Search } from '@/app/ui/search';
 


interface HomeProps {
  searchParams?: {
    search?: string;
  }
}

export default async function Home({ searchParams }: Readonly<HomeProps>) {
  const { data: restaurants } = await getRestaurants();
  let restaurantsMapped = restaurants.map((restaurant) => {
    return {
      id: restaurant.id,
      name: restaurant.name,
      address: restaurant.address,
      photoSrc: restaurant.image,
      isOpen: isValidHours(restaurant.hours)
    }
  })

  const query = searchParams?.search ?? '';
 
  if(query) {
    restaurantsMapped = restaurantsMapped.filter((restaurant) => {
      return restaurant.name.toLowerCase().includes(query.toLowerCase())
    })
  }

  return (
    <>
      <h1 className='text-2xl text-center text-gray-800 mt-8'>Bem-vindo ao Lista Rango</h1>
      <Search />
      <div className='grid grid-cols-3 gap-9' role="list">
        {
          restaurantsMapped.map((restaurant) => {
            return (
              <CardRestaurant
                id={restaurant.id}
                key={'restaurant-card-' + restaurant.id}
                name={restaurant.name}
                address={restaurant.address}
                isOpen={restaurant.isOpen}
                photoSrc={restaurant.photoSrc}
              />
            )
          })
        }
      </div>
    </>
  )
}
