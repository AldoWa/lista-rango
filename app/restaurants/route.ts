import { NextResponse } from "next/server";
import { Days } from "../libs/types";

export async function GET() {
  return NextResponse.json({
    data: [
      {
        id: 1,
        name: 'Restaurant 1',
        address: 'Address 1',
        image: 'https://via.placeholder.com/112x112',
        hours: [
          {
            "from": '10:00',
            "to": '22:00',
            "days": [Days.DOMINGO,Days.QUINTA,Days.SEXTA]
          }
        ]
      },
      {
        id: 2,
        name: 'Restaurant 2',
        address: 'Address 2',
        image: 'https://via.placeholder.com/112x112',
        hours: [
          {
            "from": '15:00',
            "to": '22:00',
            "days": [Days.DOMINGO, Days.SEGUNDA, Days.TERCA, Days.QUARTA, Days.QUINTA, Days.SEXTA]
          }
        ]
      },
      {
        id: '3',
        name: 'Restaurant 3',
        address: 'Address 3',
        image: 'https://via.placeholder.com/112x112',
        hours: [
          {
            "from": '17:00',
            "to": '01:00',
            "days": [Days.DOMINGO, Days.SEGUNDA, Days.QUINTA, Days.SEXTA]
          }
        ]
      },
    ]
  })
}