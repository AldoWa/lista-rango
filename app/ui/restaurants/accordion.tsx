'use client'

import React, { useState } from "react";
import ChevronDown from "@/public/chevron-down.svg";
import Image from "next/image";

interface AccordionProps {
  title: string;
  children: React.ReactNode;
}
export function Accordion({ title, children }: Readonly<AccordionProps>) {
  const [isOpen, setIsOpen] = useState(false);

  function openAccordion() {
    setIsOpen(opened => !opened)
  }

  return (
    <div data-testid="accordion">
      <h2 id="accordion-flush-heading-1" className="px-2.5">
        <button
          type="button"
          className="flex items-center justify-between w-full py-5 font-medium rtl:text-right text-gray-500 border-b border-b-neutral-950 gap-3"
          aria-expanded="true"
          aria-controls="accordion-flush-body-1"
          onClick={openAccordion}
          data-testid="accordion-button"
        >
          <span className="text-slate-950 ml-3" data-testid="accordion-title">{ title }</span>
          <Image
            src={ChevronDown}
            alt="Open accordion"
            height={10}
            width={16}
            className={`transition ease-in-out w-3 h-3 ${isOpen ? 'rotate-0' : '-rotate-90'} shrink-0 text-slate-950 mr-3`}
          ></Image>
        </button>
      </h2>

      <div id="accordion-flush-body-1" className={isOpen ? 'block' : 'hidden'} aria-labelledby="accordion-flush-heading-1" data-testid="accordion-body">
        <div className="mt-6 grid grid-cols-2 gap-8">
          { children }
        </div>
      </div>
    </div>
  );
}
