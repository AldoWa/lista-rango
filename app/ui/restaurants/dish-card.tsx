import PratoRestaurant from "@/public/prato-de-restaurante-vegetariano@2x.png";
import AwardSvg from "@/public/award.svg";

import Image from "next/image";

interface DishCardProps {
  onClickOpenModal: () => void;
  title: string;
  price: string;
  hasPromo: boolean;
}

export function DishCard({ onClickOpenModal, title, price, hasPromo }: DishCardProps) {
  return (
    <button
      type="button"
      className="shadow-md flex flex-row max-[600px]:flex-col max-[600px]:items-center"
      aria-live="polite"
      onClick={onClickOpenModal}
    >
      <Image
        src={PratoRestaurant}
        width={116}
        height={116}
        alt="Prato Restaurant"
      />
      <div className="flex flex-col gap-y-2 p-4 text-slate-950 items-start">
        <div className="flex items-center w-full items-start max-[600px]:flex-col">
          <h2 className="text-base font-medium">{title}</h2>
          { hasPromo && (
            <span
              role="alert"
              aria-live="polite"
              className="flex bg-violet-800 py-1.5 px-4 rounded-xl max-h-5 flex items-center gap-x-1"
            >
              <Image src={AwardSvg} alt="award" width={9} height={12} />
              <span className="text-white text-xs font-bold">Promo Almo√ßo</span>
            </span>
          )}
        </div>
        <p className="text-slate-950 font-normal text-sm text-left">
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
        </p>
        <span className="text-cyan-700 text-base font-normal">{price}</span>
      </div>
    </button>
  );
}
