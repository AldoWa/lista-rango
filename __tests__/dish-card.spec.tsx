import { DishCard } from "@/app/ui/restaurants/dish-card"
import { fireEvent, render, screen } from "@testing-library/react"

describe('dish-card', () => {
  it('Should render dish-card with promo off', () => {
    render(
      <DishCard
        hasPromo={false}
        onClickOpenModal={() => {}}
        price="R$ 20,00"
        title="Prato"
      ></DishCard>
    )
    
    const dishCardPrice = screen.getByTestId('dish-card-price')
    const dishCardTitle = screen.getByTestId('dish-card-title')
    const dishCardPromo = screen.queryByTestId('dish-card-promo')
    const dishCardButton = screen.getByTestId('dish-card-button')

    expect(dishCardButton).toBeInTheDocument()
    expect(dishCardPromo).not.toBeInTheDocument()
    expect(dishCardTitle).toBeInTheDocument()
    expect(dishCardPrice).toBeInTheDocument()

    expect(dishCardTitle).toHaveTextContent('Prato')
    expect(dishCardPrice).toHaveTextContent('R$ 20,00')
  })

  it('Should render dish-card with promo on', () => {
    render(
      <DishCard
        hasPromo
        onClickOpenModal={() => {}}
        price="R$ 20,00"
        title="Prato"
      ></DishCard>
    )

    
    const dishCardPromo = screen.queryByTestId('dish-card-promo')

    expect(dishCardPromo).toBeInTheDocument()
    expect(dishCardPromo).toHaveTextContent('Promo AlmoÃ§o')
  })

  it('Should click on the button', () => {
    const onClickOpenModal = jest.fn()
    render(
      <DishCard
        hasPromo
        onClickOpenModal={onClickOpenModal}
        price="R$ 20,00"
        title="Prato"
      ></DishCard>
    )

    const dishCardButton = screen.getByTestId('dish-card-button')
    fireEvent.click(dishCardButton)

    expect(onClickOpenModal).toHaveBeenCalled()
  })

  it('Should match snapshot', async () => {
    const { container } = render(
      <DishCard
        hasPromo={false}
        onClickOpenModal={() => {}}
        price="R$ 20,00"
        title="Prato"
      ></DishCard>
    )
    expect(container).toMatchSnapshot()
  })
})