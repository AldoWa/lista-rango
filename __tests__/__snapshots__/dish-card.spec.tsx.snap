// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dish-card Should match snapshot 1`] = `
<div>
  <button
    class="shadow-md flex flex-row max-[600px]:flex-col max-[600px]:items-center"
    data-testid="dish-card-button"
    type="button"
  >
    <img
      alt="Prato Restaurant"
      data-nimg="1"
      decoding="async"
      height="116"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=256&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=128&q=75 1x, /_next/image?url=%2Fimg.jpg&w=256&q=75 2x"
      style="color: transparent;"
      width="116"
    />
    <div
      class="flex flex-col gap-y-2 p-4 pb-0 pt-1 text-slate-950 items-start"
    >
      <div
        class="flex items-center w-full items-start max-[600px]:flex-col justify-between "
      >
        <h2
          class="text-base font-medium"
          data-testid="dish-card-title"
        >
          Prato
        </h2>
      </div>
      <p
        class="text-slate-950 font-normal text-sm text-left"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
      </p>
      <span
        class="text-cyan-700 text-base font-normal"
        data-testid="dish-card-price"
      >
        R$ 20,00
      </span>
    </div>
  </button>
</div>
`;
